examples/lilygo-t5-47-epd-platformio/src/main.cpp

```diff
@@ -0,0 +1,243 @@
+#include <Arduino.h>
+
+// esp32 sdk imports
+#include "esp_heap_caps.h"
+#include "esp_log.h"
+
+// epd
+#include "epd_driver.h"
+#include "epd_highlevel.h"
+
+// battery
+#include <driver/adc.h>
+#include "esp_adc_cal.h"
+
+// deepsleep
+#include "esp_sleep.h"
+
+// font
+#include "Firasans.h"
+
+#define BATT_PIN 36
+
+#define WAVEFORM EPD_BUILTIN_WAVEFORM
+
+/**
+ * Upper most button on side of device. Used to setup as wakeup source to start from deepsleep.
+ * Pinout here https://ae01.alicdn.com/kf/H133488d889bd4dd4942fbc1415e0deb1j.jpg
+ */
+gpio_num_t FIRST_BTN_PIN = GPIO_NUM_39;
+
+EpdiyHighlevelState hl;
+// epd_ambient_temperature() exists
+// but did not work for this device in my testing
+// so hardcode to about your room temp
+int temperature = 20;
+uint8_t *fb;
+enum EpdDrawError err;
+
+// CHOOSE HERE YOU IF YOU WANT PORTRAIT OR LANDSCAPE
+// both orientations possible
+EpdRotation orientation = EPD_ROT_PORTRAIT;
+// EpdRotation orientation = EPD_ROT_LANDSCAPE;
+
+int vref = 1100;
+
+/**
+ * RTC Memory var to get number of wakeups via wakeup source button
+ * For demo purposes of rtc data attr
+**/
+RTC_DATA_ATTR int pressed_wakeup_btn_index;
+
+/**
+ * That's maximum 30 seconds of runtime in microseconds
+ */
+int64_t maxTimeRunning = 30000000;
+
+
+double_t get_battery_percentage()
+{
+    // When reading the battery voltage, POWER_EN must be turned on
+    epd_poweron();
+    delay(50);
+
+    Serial.println(epd_ambient_temperature());
+
+    uint16_t v = analogRead(BATT_PIN);
+    Serial.print("Battery analogRead value is");
+    Serial.println(v);
+    double_t battery_voltage = ((double_t)v / 4095.0) * 2.0 * 3.3 * (vref / 1000.0);
+
+    // Better formula needed I suppose
+    // experimental super simple percent estimate no lookup anything just divide by 100
+    double_t percent_experiment = ((battery_voltage - 3.7) / 0.5) * 100;
+
+    // cap out battery at 100%
+    // on charging it spikes higher
+    if (percent_experiment > 100) {
+        percent_experiment = 100;
+    }
+
+    String voltage = "Battery Voltage :" + String(battery_voltage) + "V which is around " + String(percent_experiment) + "%";
+    Serial.println(voltage);
+
+    epd_poweroff();
+    delay(50);
+
+    return percent_experiment;
+}
+
+void display_center_message(const char* text) {
+    // first clear screen
+    epd_poweron();
+    epd_clear();
+    err = epd_hl_update_screen(&hl, MODE_GL16, temperature);
+    epd_poweroff();
+    delay(100);
```

Comment:
For me, this produces rendering artifacts, because clear() is used without clearing the framebuffer.
I would clear the screen after power on, and then use the highlevel API:
`
void display_center_message(const char* text) {

    epd_hl_set_all_white(&hl);

    int cursor_x = EPD_WIDTH / 2;
    int cursor_y = EPD_HEIGHT / 2;
    if (orientation == EPD_ROT_PORTRAIT) {
        // height and width switched here because portrait mode
        cursor_x = EPD_HEIGHT / 2;
        cursor_y = EPD_WIDTH / 2;
    }
    EpdFontProperties font_props = epd_font_properties_default();
    font_props.flags = EPD_DRAW_ALIGN_CENTER;
    epd_write_string(&FiraSans_12, text, &cursor_x, &cursor_y, fb, &font_props);

    epd_poweron();
    err = epd_hl_update_screen(&hl, MODE_GC16, temperature);
    delay(500);
    epd_poweroff();
    delay(1000);
}
`
Also, the delays seem unnecessary for me?

Reply:
I now use the `epd_hl_set_all_white` method, but the delays reduce some artifacts or graying of the background in my testing.
Maybe they can be avoided with further changes to the driving code here